lib/Sema/TypeCheckExpr.cpp

```diff
@@ -655,116 +656,78 @@ static Type lookupDefaultLiteralType(TypeChecker &TC, DeclContext *dc,
-  if (!type)
+static std::pair<const char *, bool>
+getDefaultTypeNameAndPerformLocalLookup(const ProtocolDecl *protocol) {
+  TypeChecker &tc = TypeChecker::createForContext(protocol->getASTContext());
+
+#define TYPE_NAME_CASE(KIND, NAME, PERFORM_LOCAL_LOOKUP)                       \
+  if (protocol == tc.getProtocol(SourceLoc(), KnownProtocolKind::KIND))        \
+    return std::make_pair(NAME, PERFORM_LOCAL_LOOKUP);
+
+  TYPE_NAME_CASE(ExpressibleByUnicodeScalarLiteral, "UnicodeScalarType", true)
+  TYPE_NAME_CASE(ExpressibleByExtendedGraphemeClusterLiteral,
+                 "ExtendedGraphemeClusterType", true)
+  TYPE_NAME_CASE(ExpressibleByStringLiteral, "StringLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByStringInterpolation, "StringLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByIntegerLiteral, "IntegerLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByFloatLiteral, "FloatLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByBooleanLiteral, "BooleanLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByArrayLiteral, "Array", false)
+  TYPE_NAME_CASE(ExpressibleByDictionaryLiteral, "Dictionary", false)
+  TYPE_NAME_CASE(ExpressibleByColorLiteral, "_ColorLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByImageLiteral, "_ImageLiteralType", true)
+  TYPE_NAME_CASE(ExpressibleByFileReferenceLiteral, "_FileReferenceLiteralType",
+                 true)
+
+#undef DEFAULT_TYPE_NAME_CASE
+
+  return std::make_pair(nullptr, false);
+}
+
+llvm::Expected<Type>
+swift::DefaultTypeRequest::evaluate(Evaluator &evaluator, const char *name,
+                                    const bool performLocalLookup,
+                                    const DeclContextWrapper x) const {
+  if (!name)
     return nullptr;
 
-  // If we haven't found the type yet, look for it now.
-  if (!*type) {
-    if (performLocalLookup)
-      *type = lookupDefaultLiteralType(*this, dc, name);
+  const DeclContext *dc = x.dc;
+  TypeChecker &tc = getTypeChecker();
 
-    if (!*type)
-      *type = lookupDefaultLiteralType(*this, getStdlibModule(dc), name);
+  Type type;
+  if (performLocalLookup)
+    type = lookupDefaultLiteralType(tc, dc, name);
 
-    // Strip off one level of sugar; we don't actually want to print
-    // the name of the typealias itself anywhere.
-    if (type && *type) {
-      if (auto boundTypeAlias =
-                 dyn_cast<TypeAliasType>(type->getPointer()))
-        *type = boundTypeAlias->getSinglyDesugaredType();
-    }
+  if (!type)
+    type = lookupDefaultLiteralType(tc, tc.getStdlibModule(dc), name);
+
+  // Strip off one level of sugar; we don't actually want to print
+  // the name of the typealias itself anywhere.
+  if (type) {
+    if (auto boundTypeAlias = dyn_cast<TypeAliasType>(type.getPointer()))
+      type = boundTypeAlias->getSinglyDesugaredType();
   }
+  return type;
+}
 
-  return *type;
+TypeChecker &DefaultTypeRequest::getTypeChecker() const {
+  return TypeChecker::createForContext(getDeclContext()->getASTContext());
```

Comment:
Rather than creating a new type checker here, use

```
*static_cast<TypeChecker *>(getDeclContext()->getASTContext().getLazyResolver())
```

... which is ugly, but always works because the type checker lives forever. I was sorta hoping that  we could eliminate the need for the `TypeChecker` to be passed through this request at all, but we can get there later.

Reply:
Oh, I retract this. `TypeChecker::createForContext` isn't what I thought it was, sorry!